src/event-scope.ts

```diff
@@ -1,177 +1,191 @@
+				if (exitSubscriptions && exitSubscriptions.length > 0) {
+					// If there is no parent scope, then go ahead and execute the 'exit' event
+					if (this.parent === null || !this.parent.isActive) {
+						// Record the initial "version" before starting to call subscribers
+						this._exitEventVersion = 0;
+
+						// Invoke all subscribers
+						(this._onExit as Event<EventScope, EventScopeExitEventArgs>).publish(this, { abort: false });
+
+						// Delete the field to indicate that raising the exit event suceeded
+						delete this._exitEventVersion;
 					}
-
-					// Move subscribers to the parent scope
-					exitSubscriptions.forEach(sub => {
-						if (!sub.isOnce || !sub.isExecuted) {
-							this.parent.onExit.subscribe(sub.handler);
+					else {
+						try {
+							// Attempt to move subscribers to the parent scope
+							this.parent.receiveEventHandlers(exitSubscriptions);
+						}
+						catch (e) {
+							this.dispose({ abort: true });
+							throw e;
 						}
-					});
-
-					if (this.parent._exitEventVersion !== undefined) {
-						this.parent._exitEventVersion++;
 					}
 				}
-
-				// Clear the events to ensure that they aren't
-				// inadvertantly raised again through this scope
-				this.onAbort.clear();
-				this.onExit.clear();
 			}
+
+			// Clear the events to ensure that they aren't inadvertantly raised again through this scope
+			this._onExit.clear();
 		}
 		finally {
 			// The event scope is no longer active
 			this.isActive = false;
-
-			if (EventScope$current && EventScope$current === this) {
-				// Roll back to the closest active scope
-				while (EventScope$current && !EventScope$current.isActive) {
-					EventScope$current = EventScope$current.parent;
-				}
-			}
-		}
-	}
-}
-
-export function EventScope$onExit(callback: Function, thisPtr: any = null): void {
-	if (EventScope$current === null) {
-		// Immediately invoke the callback
-		if (thisPtr) {
-			callback.call(thisPtr);
-		}
-		else {
-			callback();
 		}
 	}
-	else if (!EventScope$current.isActive) {
-		throw new Error("The current event scope cannot be inactive.");
-	}
-	else {
-		// Subscribe to the exit event
-		EventScope$current.onExit.subscribe(callback.bind(thisPtr));
-	}
-}
 
-export function EventScope$onAbort(callback: Function, thisPtr: any = null): void {
-	if (EventScope$current !== null) {
-		if (!EventScope$current.isActive) {
-			throw new Error("The current event scope cannot be inactive.");
+	private receiveEventHandlers(subscriptions: EventSubscription<EventScope, EventScopeExitEventArgs>[]) {
+		var maxNesting = this.settings.maxExitingTransferCount - 1;
+		if (this._exitEventVersion >= maxNesting) {
+			throw new Error("Exceeded max scope event transfer.");
 		}
 
-		// Subscribe to the abort event
-		EventScope$current.onAbort.subscribe(callback.bind(thisPtr));
-	}
-}
+		// Move subscribers to the parent scope
+		subscriptions.forEach(sub => {
+			if (!sub.isOnce || !sub.isExecuted) {
```

Comment:
Comment as to why `isOnce` subscriptions aren't transferred?

Reply:
Ha, good question! I believe that's existing code/logic, so I'll dig into it to see if I can figure out, or remember why.

Reply:
Oh, right. "One time" subscriptions aren't transferred if they've already been run. This wasn't needed in Exo because the functor was added directly to the event subscription and it internally tracked this, whereas here we're directly subscribing the event handler. However, there is a logical error here in that if it is a one-time subscription that hasn't been run it will be subscribed as not a one-time subscription. But, it doesn't look like the one-time subscription concept is used (at least internally), so maybe I can just get rid of it. I may just tweak this logic though, since that's probably going to be simpler (and safer) than ripping out a bunch of code.

Reply:
I went ahead and just simplified it since the "filter" and "once" options aren't being leveraged